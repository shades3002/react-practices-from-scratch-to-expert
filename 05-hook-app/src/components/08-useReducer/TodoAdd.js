import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1 ) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddTodo( newTodo );
        reset();
    };

    return (
        <>
            <h4>Add Todo</h4>
            <hr />

            <form onSubmit={ handleSubmit }>
                <input 
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Todo ..."
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />
                <div className="d-grid gap-2">
                    <button
                        type="submit"
                        className="btn btn-outline-primary mt-1 btn-block"
                    >
                        Add
                    </button>
                </div>
            </form>
        </>
    )
}
