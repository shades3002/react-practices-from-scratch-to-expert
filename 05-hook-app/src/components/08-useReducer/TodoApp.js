import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';

import './style.css';

const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    });

    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify( todos ));
    }, [ todos ]);

    const handleDelete = ( id ) => {

        const deleteTodoAction = {
            type: 'delete',
            payload: id
        };

        dispatch( deleteTodoAction );
    }

    const handleToggle = ( id ) => {
        dispatch({
            type: 'toggle',
            payload: id
        });
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1 ) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        const addTodoAction = {
            type: 'add',
            payload: newTodo
        };

        dispatch( addTodoAction );
        reset();
    }

    return (
        <div>
            <h1>TodoApp ({ todos.length }) </h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush ">
                        {
                            todos.map( (todo, i)  => (
                                <li
                                    key={ todo.id }
                                    className="list-group-item w-100"
                                > 
                                    <p 
                                        className={ `w-50 ${ todo.done && 'complete' }`}
                                        onClick={ () => handleToggle( todo.id ) }
                                    >
                                        { i + 1}. { todo.desc } 
                                    </p>
                                    <div className="d-grid gap-2 d-md-flex w-50 justify-content-md-start">
                                        <button
                                            className="btn btn-danger ms-2"
                                            onClick={ () => handleDelete(todo.id) }
                                        >
                                            Delete
                                        </button>
                                    </div>
                                </li> 
                            ))
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Add Todo</h4>
                    <hr />

                    <form onSubmit={ handleSubmit }>
                        <input 
                            type="text"
                            name="description"
                            className="form-control"
                            placeholder="Todo ..."
                            autoComplete="off"
                            value={ description }
                            onChange={ handleInputChange }
                        />
                        <div className="d-grid gap-2">
                            <button
                                type="submit"
                                className="btn btn-outline-primary mt-1 btn-block"
                            >
                                Add
                            </button>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}
