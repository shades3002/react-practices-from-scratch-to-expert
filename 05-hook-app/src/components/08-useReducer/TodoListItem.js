import React from 'react'

export const TodoListItem = ({ todo, index, handleToggle, handleDelete }) => {
    return (
        <li
            className="list-group-item w-100"
        > 
            <p 
                className={ `w-50 ${ todo.done && 'complete' }`}
                onClick={ () => handleToggle( todo.id ) }
            >
                { index + 1}. { todo.desc } 
            </p>
            <div className="d-grid gap-2 d-md-flex w-50 justify-content-md-start">
                <button
                    className="btn btn-danger ms-2"
                    onClick={ () => handleDelete(todo.id) }
                >
                    Delete
                </button>
            </div>
        </li> 
    )
}
