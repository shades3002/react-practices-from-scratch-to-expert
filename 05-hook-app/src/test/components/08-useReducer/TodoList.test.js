import { shallow } from 'enzyme';
import React from 'react';
import { TodoList } from "../../../components/08-useReducer/TodoList";
import { demoTodos } from '../fixtures/demoTodos';


describe('test in <TodoList />', () => {
    
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow( 
        <TodoList 
            todos={ demoTodos }
            handleDelete={ handleDelete } 
            handleToggle={ handleToggle }
        
        />
    );

    test('should show correctly', () => {
        
        expect( wrapper ).toMatchSnapshot();
    });

    test('should have two <TodoListItem />', () => {
        expect( wrapper.find('TodoListItem').length ).toBe( demoTodos.length );
        expect( wrapper.find('TodoListItem').at(0).prop('handleDelete') ).toEqual( expect.any(Function));
    });
    
});
