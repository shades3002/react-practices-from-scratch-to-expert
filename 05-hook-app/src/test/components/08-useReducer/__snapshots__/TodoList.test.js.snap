// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <TodoList /> should show correctly 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "test 1",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "test 2",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
